systembolaget/stock.go

```diff
@@ -0,0 +1,84 @@
+package systembolaget
+
+import (
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log/slog"
+	"net/http"
+	"net/url"
+)
+
+type StockBalance struct {
+	ProductNumber string `json:"productId"`
+	StoreID       string `json:"storeId"`
+	StockLevel    int    `json:"stock"`
+	Shelf         string `json:"shelf"`
+}
+
+func (c *Client) StockBalance(ctx context.Context, storeID string, productNumber string) (*StockBalance, error) {
+	log := GetLogger(ctx)
+
+	u := &url.URL{
+		Scheme: "https",
+		Host:   "api-extern.systembolaget.se",
+		Path:   fmt.Sprintf("/sb-api-ecommerce/v1/stockbalance/store/%s/%s/", storeID, productNumber),
+	}
+
+	log = log.With(slog.String("url", u.String()))
+
+	header := http.Header{}
+	header.Set("Origin", "https://www.systembolaget.se")
+	header.Set("Access-Control-Allow-Origin", "*")
+	header.Set("Pragma", "no-cache")
+	header.Set("Accept", "application/json")
+	header.Set("Accept-Encoding", "gzip")
+	header.Set("Cache-Control", "no-cache")
+	header.Set("Ocp-Apim-Subscription-Key", c.apiKey)
+
+	if c.userAgent != "" {
+		header.Set("User-Agent", c.userAgent)
+	}
+
+	req := (&http.Request{
+		Method: http.MethodGet,
+		URL:    u,
+		Header: header,
+	}).Clone(ctx)
+
+	log.Debug("Performing request")
+	res, err := c.httpClient.Do(req)
+	if err != nil {
+		log.Error("Request failed", slog.Any("error", err))
+		return nil, err
+	}
+
+	if res.StatusCode != http.StatusOK {
```

Comment:
Interesting that the status code seems to be 200 in the error case as well. Here's the body in the error case:

```json
{"productId":null,"storeId":null,"shelf":null,"stock":null,"isInStoreAssortment":false}
```

I guess one solution is to have an internal stockBalance type that more closely follows the API;

```go
type stockBalance struct {
  	ProductNumber *string `json:"productId"`
 	StoreID       *string `json:"storeId"`
 	StockLevel    int    `json:"stock"`
 	Shelf         *string `json:"shelf"`
}
```

And then check if shelf etc. is nil, if it is, return some error. Otherwise, map the values onto the current `StockBalance` type.